internal/grpc/services/storageprovider/storageprovider.go

```diff
@@ -1241,16 +1286,27 @@ func (v descendingMtime) Swap(i, j int) {
 	v[i], v[j] = v[j], v[i]
 }
 
-func unwrapProviderID(id *provider.ResourceId) string {
+func unwrapProviderIDAndPath(ref *provider.Reference, mountPath string) string {
 	var spid string
-	if id != nil {
-		spid, id.StorageId = storagespace.SplitStorageID(id.StorageId)
+	if ref.ResourceId != nil {
+		spid, ref.ResourceId.StorageId = storagespace.SplitStorageID(ref.ResourceId.StorageId)
 	}
+
+	// Trim the mount path of the storage provider if the request was for an absolute reference
+	if ref != nil && !utils.IsRelativeReference(ref) {
+		ref.Path = strings.TrimPrefix(ref.Path, mountPath)
+	}
+
 	return spid
 }
 
-func rewrapProviderID(id *provider.ResourceId, spid string) {
-	if id != nil {
-		id.StorageId = storagespace.FormatStorageID(spid, id.StorageId)
+func rewrapProviderIDAndPath(ref *provider.Reference, spid, mountPath string) {
+	if ref.ResourceId != nil {
+		ref.ResourceId.StorageId = storagespace.FormatStorageID(spid, ref.ResourceId.StorageId)
+	}
+
+	// Prepend the mount path of the storage provider if the request was for an absolute reference
+	if ref != nil && !utils.IsRelativeReference(ref) {
+		ref.Path = path.Join(mountPath, ref.Path)
```

Comment:
This is just my current braindump, I need to think about this further ...

Reply:
> But this path must not be exposed to prevent share receivers from getting to know how someoneelse organized his files

But this is exactly what we want to do. Just leave it empty and it'll work how you want it to. I could move this to the EOS driver but this fits better in the storageprovider layer IMO

Reply:
Hm, as a feature it makes sense to have a way to enable disclosing the paths 🤔

Reply:
@ishank011 If all you want to achive is being able to shard the user directory, why can you not implement that logic in the registry, with the gateway doing the path wrapping and unwrapping? Is the problem that the storage provider will then only see a relative path inside a space, eg:
- gateway receives `/users/e/einstein/some/path`
- registry matches `/users/e` with the `/users/[a-k]` mount_point regex
- registry cuts off the matched prefix for the provider `/users/e` and returns a provider with the mount_point set to `/users/e`
- gateway cuts off the mount_point from the path, which becomes `einstein/some/path`
- storage driver has to looks up the space for `einstein`, lets call it the `space_path`
- the driver can navigate to `some/path`, relative to einsteins home

now this does not involve any id lookups ... this would be purely path based resource access ... without the storage provider having to know the mount_point?

At this point the storage provider still has to know about the space_path, maybe using a template like `{.Spaceid}` or `{.Owner.Id}`. Which is why the space registy can be configured with a `path_template`. It serves the same purpose, but now all path logic is in the registry.

IMO the code complexity would decrease if we use the path / storageproviderid to let the registry look up the storage space once and then always continue with relative references.

We should have a call to clarify the how to proceed ...